hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java

```diff
@@ -0,0 +1,537 @@
+import org.apache.hudi.table.HoodieTimelineArchiveLog;
+import org.apache.hudi.table.UserDefinedBulkInsertPartitioner;
+import org.apache.hudi.table.action.HoodieWriteMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * @author xianghu.wang
+ * @time 2020/6/10
+ * @description
+ */
+public abstract class AbstractHoodieWriteClient<T extends HoodieRecordPayload<T>, I, K, O, P> extends AbstractHoodieClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractHoodieWriteClient.class);
+  private static final long serialVersionUID = 1L;
+  private final transient HoodieMetrics metrics;
+  private final transient AbstractHoodieIndex<T, I, K, O, P> index;
+  private final transient AbstractHoodieTable<T, I, K, O, P> table;
+  private transient Timer.Context writeContext = null;
+  private transient WriteOperationType operationType;
+
+  private static final String LOOKUP_STR = "lookup";
+  private final boolean rollbackPending;
+  private transient Timer.Context compactionTimer;
+
+
+  public void setOperationType(WriteOperationType operationType) {
+    this.operationType = operationType;
+  }
+
+  public WriteOperationType getOperationType() {
+    return this.operationType;
+  }
+
+  public AbstractHoodieWriteClient(HoodieEngineContext context, AbstractHoodieIndex<T, I, K, O, P> index,
+                                   AbstractHoodieTable<T, I, K, O, P> table,HoodieWriteConfig clientConfig) {
+    this(context, index, table,clientConfig, Option.empty());
+  }
+
+  protected AbstractHoodieWriteClient(HoodieEngineContext context, AbstractHoodieIndex<T, I, K, O, P> index,
+                                      AbstractHoodieTable<T, I, K, O, P> table,HoodieWriteConfig clientConfig,
+                                      Option<AbstractEmbeddedTimelineService> timelineServer) {
+    this(context, index, table,clientConfig, timelineServer, false);
+  }
+
+  public AbstractHoodieWriteClient(HoodieEngineContext context, AbstractHoodieIndex<T, I, K, O, P> index,AbstractHoodieTable<T, I, K, O, P> table,
+                                   HoodieWriteConfig clientConfig, Option<AbstractEmbeddedTimelineService> timelineServer, boolean rollbackPending) {
+    super(context, clientConfig, timelineServer);
+    this.index = index;
+    this.table = table;
+    this.metrics = new HoodieMetrics(config, config.getTableName());
+    this.rollbackPending = rollbackPending;
+  }
+
+  public abstract I filterExists(I hoodieRecords);
+
+  public abstract O upsert(I records, final String instantTime);
+
+  public abstract O upsertPreppedRecords(I preppedRecords, final String instantTime);
+
+  public abstract O insert(I records, final String instantTime);
+
+  public abstract O insertPreppedRecords(I preppedRecords, final String instantTime);
```

Comment:
these are methods from HoodieWriteClient, and users would use HoodieWriteClient to upsert/insert records directly using the APIs, right now the HoodieWriteClient has been removed, so it breaks the compatibility.

Reply:
> these are methods from HoodieWriteClient, and users would use HoodieWriteClient to upsert/insert records directly using the APIs, right now the HoodieWriteClient has been removed, so it breaks the compatibility.

@leesf Yes, it is not finished yet. I have noticed that HoodieWriteClient has been referenced in many places(eg hudi-cli,hudi-utilities...). When hudi-client module is ready, the other modules which rely on hudi-client should make appropriate changes to adapt to.